[docs][charts] Reorganize the Tooltip documentation (#17917)
mui/mui-x: docs/data/charts/tooltip/tooltip.md

/>
```

### Series

To hide a series, the formatted value should be `null`.

> To display en empty space, return an empty string.

Author: ```suggestion
To display the series with a blank space, return an empty string.
```

## Style modification

The tooltip can be styled using CSS classes, similar to other elements.

> However, there is one caveat du to the usage of [Portal](https://react.dev/reference/react-dom/createPortal).

Author: ```suggestion
However, there is one caveat regarding using [portal](https://react.dev/reference/react-dom/createPortal).
```

> The tooltip renders as a children of the document's body element.

Author: ```suggestion
The tooltip renders as a child of the document's body element.
From a DOM perspective, it's not inside the chart.
```

So using the chart's `sx` prop as follow does not work.

```tsx
import { chartsTooltipClasses } from '@mui/x-charts';

<LineChart
  sx={{
    [`& .${chartsTooltipClasses.root} .${chartsTooltipClasses.valueCell}`]: {
      color: 'red',
    },
  }}
/>;
```

> To apply the same style as we're trying to apply above, we need to use the `sx` property inside the tooltip itself thanks to `slotProps.tooltip`:

Author: ```suggestion
To apply the same style as above, use the `sx` prop of the tooltip itself, thanks to `slotProps.tooltip`.
```

{{"demo": "TooltipStyle.js"}}

> If necessary, you can also disable the portal by setting `slotProps.tooltip.disablePortal` to `true`.

Author: ```suggestion
Another option is to disable the portal by setting `slotProps.tooltip.disablePortal` to `true`.
```

> In that case the tooltip will render as a child of the chart.

Author: ```suggestion
In that case, the tooltip renders as a child of the chart, and CSS rules apply as expected.
```

## Using a custom tooltip

> For advanced use-case, it can be necessary to create your own tooltip.

Author: ```suggestion
For advanced use cases, it can be necessary to create your own tooltip.
```

You can replace the default tooltip in single component charts by using slots.

```jsx
<LineChart slots={{ tooltip: CustomItemTooltip }} />
```

With compositon, you can use your component inside the container.

```jsx
<ChartContainer>
  // ...
  <CustomItemTooltip />
</ChartContainer>
```

:::warning
If your custom tooltip does not use portal, it can not render inside the ChartContainer.
Overwise it would render an HTML element inside a SVG.

The solution is to put your tooltip inside the ChartDataProvider to get data, but outside ChartSurface to avoid being inside the SVG.

```jsx
<ChartDataProvider>
  <ChartSurface>{/* ... */}</ChartSurface>
  <CustomItemTooltip disablePortal />
</ChartDataProvider>
```

:::

## Creating a tooltip

> To create your custom tooltip, we provide some helpers which are explained in following sections:

Author: ```suggestion
To create your custom tooltip, the library exports some helpers which are explained in the following sections:
```

- `<ChartsTooltipContainer />` A wrapper providing the open/close state and the position of the tooltip.
- `<ChartsItemTooltipContent />` render the content of the default item tooltip.
- `<ChartsAxisTooltipContent />` render the content of the default axis tooltip.
- `useItemTooltip()` provides all basic information associated to the current item tooltip.
- `useAxesTooltip()` provides all basic information associated to the current axes tooltip.

### Modifying the position

> To override the tooltip position, you can create a wrapper that will manage the position.

Author: ```suggestion
To override the tooltip position, you can create a wrapper that manages the position.
```

```jsx
function CustomTooltipPopper(props){
  // ... (event management) ...

  return <NoSsr>
      <Popper {/* position */}>
        {props.children}
      </Popper>
    </NoSsr>
}
```

> Then you will be able to either render a built-in content (`<ChartsItemTooltipContent />` or `<ChartsAxisTooltipContent />`) or your own component.

Author: ```suggestion
Then you can either render built-in content (with `<ChartsItemTooltipContent />` or `<ChartsAxisTooltipContent />`) or your own component.
```

/>
```

### Series

To hide a series, the formatted value should be `null`.
To display the series with a blank space, return an empty string.

## Style modification

The tooltip can be styled using CSS classes, similar to other elements.
However, there is one caveat regarding using [portal](https://react.dev/reference/react-dom/createPortal).

The tooltip renders as a child of the document's body element.
From a DOM perspective, it's not inside the chart.
So using the chart's `sx` prop as follow does not work.

```tsx
import { chartsTooltipClasses } from '@mui/x-charts';

<LineChart
  sx={{
    [`& .${chartsTooltipClasses.root} .${chartsTooltipClasses.valueCell}`]: {
      color: 'red',
    },
  }}
/>;
```

> To apply the same style as above, use the `sx` prop of the tooltip itself, thanks to `slotProps.tooltip`.

Reviewer: ```suggestion
To apply the same style as above, use the `sx` prop of the tooltip itself, which should be set in `slotProps.tooltip`.
```

{{"demo": "TooltipStyle.js"}}

Another option is to disable the portal by setting `slotProps.tooltip.disablePortal` to `true`.
In that case, the tooltip renders as a child of the chart, and CSS rules apply as expected.

## Using a custom tooltip

For advanced use cases, it can be necessary to create your own tooltip.
You can replace the default tooltip in single component charts by using slots.

```jsx
<LineChart slots={{ tooltip: CustomItemTooltip }} />
```

With compositon, you can use your component inside the container.

```jsx
<ChartContainer>
  // ...
  <CustomItemTooltip />
</ChartContainer>
```

:::warning
If your custom tooltip does not use portal, it can not render inside the ChartContainer.
Overwise it would render an HTML element inside a SVG.

> The solution is to put your tooltip inside the ChartDataProvider to get data, but outside ChartSurface to avoid being inside the SVG.

Reviewer: ```suggestion
The solution is to render your tooltip as a descendant the ChartDataProvider so it can access the chart data, but outside ChartSurface so it isn't rendered inside an SVG element.
```

```jsx
<ChartDataProvider>
  <ChartSurface>{/* ... */}</ChartSurface>
  <CustomItemTooltip disablePortal />
</ChartDataProvider>
```

:::

## Creating a tooltip

To create your custom tooltip, the library exports some helpers which are explained in the following sections:

> - `<ChartsTooltipContainer />` A wrapper providing the open/close state and the position of the tooltip.
> - `<ChartsItemTooltipContent />` render the content of the default item tooltip.
> - `<ChartsAxisTooltipContent />` render the content of the default axis tooltip.
> - `useItemTooltip()` provides all basic information associated to the current item tooltip.
> - `useAxesTooltip()` provides all basic information associated to the current axes tooltip.

Reviewer: ```suggestion
- `<ChartsTooltipContainer />` a wrapper providing the open/close state and the position of the tooltip.
- `<ChartsItemTooltipContent />` renders the content of the default item tooltip.
- `<ChartsAxisTooltipContent />` renders the content of the default axis tooltip.
- `useItemTooltip()` provides all basic information associated to the current item tooltip.
- `useAxesTooltip()` provides all basic information associated to the current axes tooltip.
```

### Modifying the position

To override the tooltip position, you can create a wrapper that manages the position.

```jsx
function CustomTooltipPopper(props){
  // ... (event management) ...

  return <NoSsr>
      <Popper {/* position */}>
        {props.children}
      </Popper>
    </NoSsr>
}
```

Then you can either render built-in content (with `<ChartsItemTooltipContent />` or `<ChartsAxisTooltipContent />`) or your own component.

```jsx
<CustomTooltipPopper>
  <ChartsItemTooltipContent />
</CustomTooltipPopper>
```

> The following demo shows example about how to use additional hooks such as `useXAxis()` or `useDrawingArea()` to customize the tooltip position.

Reviewer: ```suggestion
The following demo shows how to use additional hooks such as `useXAxis()` and `useDrawingArea()` to customize the tooltip position.
```

{{"demo": "CustomTooltipPosition.js"}}

#### Modifying the content

To keep the default placement, use the `<ChartsTooltipContainer />` wrapper.

> It accept a prop `trigger = 'item' | 'axis'` to know when the Popper should be open.

Reviewer: ```suggestion
It accepts a prop `trigger = 'item' | 'axis'` that defines when the Popper should open.
```

:::warning
Do not skip rendering the ChartsTooltipContainer.

> For example the following code does not work.

Reviewer: Should we briefly explain why?

Author: ```suggestion
Do not skip rendering the ChartsTooltipContainer.
Otherwise, it could miss some events triggering its opening.

For example the following code does not work.
```

Initially it's because we use the `pointerEnter` event to know if the tooltip is on desktop or mobile. But we could probably modify that and remove this warning

Reviewer: Makes sense. Here's my suggestion, but yours work too.

```suggestion
The tooltip might not open if the ChartsTooltipContainer isn't rendered. 

For example the following code does not work.
```

> But we could probably modify that and remove this warning

Yeah, that would be nice.
